### ISSUE TITLE
SoftInput (Software Keyboard) support

### ISSUE THREAD

**Issue by semtexzv** [rust-windowing/winit#1823]
On android, the application should be able to request showing & hiding of soft input.

This is available through ndk_glue for android. 

But this strangely feels like functionality that should be attached to window.
eg. for desktop we have cursor grabbing, for mobile, having methods for showing keyboard seems similar.

**Comment by torokati44** [comment]
Yeah, this would be really useful. Especially since so far I have been completely unable to pop up the virtual keyboard on `winit`'s `NativeActivity`.
I does not appear when adding`android:windowSoftInputMode="stateAlwaysVisible|adjustResize"` to the `<activity>` entry in `AndroidManifest.xml`, nor when calling `ndk_glue::native_activity().show_soft_input(true);` from Rust.

**Comment by Fancyflame** [comment]
So... does any guy have any ideas? `ndk_glue::native_activity()` is deprecated, it says using `ndk_context::android_context().vm()` instead, but what next? I'm a noob at Android programming.

**Comment by msiglreith** [comment]
Not sure what's up with the NDK function but here is an example going trough jni: https://github.com/msiglreith/pixels-android/blob/master/src/lib.rs#L30 which shows the keyboard when tapping the screen.
It uses current `main` branch as virtual keycode have just been added recently for Android, otherwise it will just return empty ):

**Comment by Fancyflame** [comment]
This example can work on my device. But do you mean this method has a compatibility problem?

**Comment by msiglreith** [comment]
Sry, I referred to the Android winit platform implementation not Android itself: https://github.com/rust-windowing/winit/pull/2226
Should work with older Android versions as well

**Comment by Fancyflame** [comment]
Thanks. That's okay if it only needs update winit to the latest version. 😃
To be honest, I cannot understand what you wrote due to my lack of knowledge about Android and Java.... Can I copy your implementation to my project? I'll mark your repo.

**Comment by Fancyflame** [comment]
@msiglreith Then, how to hide it?

**Comment by msiglreith** [comment]
> Can I copy your implementation to my project?

Feel free. I will probably look into opening a PR for this in the next days if there are no further issues with the JNI implementation.

> Then, how to hide it?

I pushed a small change in the meantime to the repository from above which implements it (hides when tapping the screen again).

**Comment by Fancyflame** [comment]
@msiglreith Thanks for your example. I'm sorry for bothering you again, but... I have another two questions: 
- How to popup and hide **safe** keyboard on Android for password input, just like `<input type="password">` does in HTML? 
- How to prevent the IME from popping up sometimes when press keys on PC (Windows 10)? (Sorry if not related to this issue)
<img width="563" alt="屏幕截图 2022-04-29 013655" src="https://user-images.githubusercontent.com/31250625/165814190-2c2b9827-5548-4388-8960-92dc7a1e08dc.png">

**Comment by msiglreith** [comment]
> How to popup and hide safe keyboard on Android for password input, just like <input type="password"> does in HTML?

I think that's more tricky and not something which can be easily implemented. I would assume creating a custom View which will set the corresponding input type but that requires interfacing/compiling Java, which would require support on the tooling front in general...

> How to prevent the IME from popping up sometimes when press keys on PC (Windows 10)? (Sorry if not related to this issue)

I haven't tried IME myself - you might want to look at https://github.com/rust-windowing/winit/pull/2243 which seems to add more functionality for it (in particular enabling/disabling IME)

**Comment by ElhamAryanpur** [comment]
are there any plans for implementation? workarounds posted here doesn't work anymore either

**Comment by madsmtm** [comment]
There's been some progress on iOS in https://github.com/rust-windowing/winit/pull/3571, and some progress on Android in https://github.com/rust-windowing/winit/pull/2993, and finally we've been having some discussion with the [Xilem folks](https://xi.zulipchat.com/) about it. I think there are vague plans to work on it in/post v0.31, but that's all I can say.

**Comment by kchibisov** [comment]
I'd add that if people are willing to put work into it, I'm all for having it. It's just should be just a little bit consistent with the rest of IME we have around, but that's about it.

**Comment by ElhamAryanpur** [comment]
I see, I'd love to help but I honestly don't know anything about JNI. will look further to see if theres some parts I can contribute to. Thank you!

**Comment by mvvvv** [comment]
I tested the Android keyboard launched via JNI on the Meta Quest2. It works exactly like any Android device.
`https://github.com/mvvvv/StereoKit-rust/blob/master/src/tools/os_api.rs#L191`

Indeed winit 0.30 blocks many characters including accented characters which it replaces completely with a backspace and erases the character typed previously. I made a  short video: https://youtu.be/os8cucfF3D8

If you need to test I have an environment for that.

**Comment by daxpedda** [comment]
Web can do this via the [`VirtualKeyboard`](https://developer.mozilla.org/en-US/docs/Web/API/VirtualKeyboard) API, only supported in Chromium.

Cc @will-hart, Winit will implement a fallback using a hidden text input as well, unless I find some better method for non-Chromium browsers.

**Comment by torokati44** [comment]
> Winit will implement a fallback using a hidden text input as well

FWIW Ruffle does the same: 
https://github.com/ruffle-rs/ruffle/blob/b81486f1dab1573861b67d322e903d5302300ceb/web/packages/core/src/internal/player/inner.tsx#L1307

**Comment by jancespivo** [comment]
I've created PR in `android-activity` https://github.com/rust-mobile/android-activity/pull/178 could you check it please?

cc @podusowski @xorgy

**Comment by mvvvv** [comment]
About this PR which is what I'm using for OpenXR/StereoKit-rust, may I suggest to enhance the KeyAction Enum with a more complete list. This will avoid the following warnings:

```
12-30 15:16:17.601 23830 23830 W ViewRootImpl[NativeActivity]: Dropping event (no window focus):KeyEvent { action=ACTION_DOWN, keyCode=KEYCODE_DEL, scanCode=0, metaState=0, flags=0x0, repeatCount=0, eventTime=0, downTime=0, deviceId=-1, source=0x0, displayId=0 }
12-30 15:16:17.601 23830 23830 W ViewRootImpl[NativeActivity]: Cancelling event (no window focus):KeyEvent { action=ACTION_UP, keyCode=KEYCODE_DEL, scanCode=0, metaState=0, flags=0x20, repeatCount=0, eventTime=0, downTime=0, deviceId=-1, source=0x0, displayId=0 }
12-30 15:16:17.601 23830 23830 W ViewRootImpl[NativeActivity]: Cancelling event (no window focus):KeyEvent { action=ACTION_UP, keyCode=KEYCODE_DEL, scanCode=0, metaState=0, flags=0x20, repeatCount=0, eventTime=0, downTime=0, deviceId=-1, source=0x0, displayId=0 }
12-30 15:16:17.602 23830 23830 W ViewRootImpl[NativeActivity]: Dropping event (no window focus):KeyEvent { action=ACTION_MULTIPLE, keyCode=KEYCODE_UNKNOWN, scanCode=0, characters="ò", metaState=0, flags=0x0, repeatCount=0, eventTime=19026369, downTime=19026369, deviceId=-1, source=0x101, displayId=-1 }
12-30 15:16:17.605 23830 23882 D SK_rust_Demos: !!!KeyboardInput: KeyEvent { physical_key: Code(Backspace), logical_key: Named(Backspace), text: None, location: Standard, state: Released, repeat: false, platform_specific: KeyEventExtra }
```

And will give access to ACTION_MULTIPLE which is the only way to get enhanced characters(here ò), smileys and utf8.

**Comment by mvvvv** [comment]
Done https://github.com/rust-windowing/winit/issues/4067
